src/net/direct.c

```diff
@@ -33,11 +33,16 @@ typedef struct qbuf {
 } *qbuf;
 
 /* return true if sendq empty */
-static boolean direct_conn_send_internal(direct_conn dc)
+static void direct_conn_send_internal(direct_conn dc)
 {
     direct_debug("dc %p\n", dc);
     list next;
 
+    /* It appears TCP_EVENT_SENT is only called from tcp_input. If in
+       the future it could ever be invoked as a result of a call to
+       tcp_write or tcp_output, this will need to be revised to avoid
+       deadlock. */
+    spin_lock(&dc->send_lock);
     while ((next = list_get_next(&dc->sendq_head))) {
```

Comment:
Locking on sendq_head seems inconsistent--direct_conn_send doesn't use the spinlock, but does disable interrupts, and here it's vice versa

Reply:
Thanks for catching that. I moved the qbuf enqueue under the locked region inside direct_conn_send_internal.